### ISSUE TITLE
CHECK routine called after syntax error

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2129]
Migrated from [rt.perl.org#3422](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3422) (status was 'resolved')

Searchable as RT3422$

**Comment by p5pRT** [comment]
### From mjtg@cus.cam.ac.uk

Playing around with one of Barrie Slaymaker's bits of code\, I tried

perl5\.6\.0 \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq "fooba"\)\, @&#8203;bar'

and got

Not enough arguments for grep at \-e line 1\, near ""fooba"\)"
\-e had compilation errors\.
unexpected OP\_GREPSTART at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 636
&nbsp;    B&#8203;::Deparse&#8203;::pp\_grepstart\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x16d498\)'\, 7\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
&nbsp;    B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x16d498\)'\, 7\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1634
&nbsp;    B&#8203;::Deparse&#8203;::pp\_list\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1810
&nbsp;    B&#8203;::Deparse&#8203;::pp\_null\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
&nbsp;    B&#8203;::Deparse&#8203;::deparse\(undef\, 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1239
&nbsp;    B&#8203;::Deparse&#8203;::deparse\_binop\_right\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1257
&nbsp;    B&#8203;::Deparse&#8203;::binop\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 6\, '='\, 7\, 1\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1303
&nbsp;    B&#8203;::Deparse&#8203;::pp\_aassign\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 6\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
&nbsp;    B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 6\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1646
&nbsp;    B&#8203;::Deparse&#8203;::pp\_list\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x16d9f0\)'\, 0\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
&nbsp;    B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x16d9f0\)'\, 0\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 706
&nbsp;    B&#8203;::Deparse&#8203;::pp\_leave\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x177a5c\)'\, 0\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
&nbsp;    B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x177a5c\)'\, 0\) called at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 375
&nbsp;    B&#8203;::Deparse&#8203;::\_\_ANON\_\_\(\) called at \(eval 4\) line 1
&nbsp;    O&#8203;::CHECK\(\) called at \-e line 1
&nbsp;    require 0 called at \-e line 1
@&#8203;x = XXX\, @&#8203;bar;

The syntax error is of course correct \(though it makes me wonder if the
"if it looks like a function \.\.\. " rule isn't overagressive sometimes\)\.

But the rest of the junk suggests that perhaps CHECK routines shouldn't
be called after syntax errors\.


Mike Guy

% perl5\.6\.0 \-V
Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.6\, archname=sun4\-solaris
&nbsp;    uname=''
&nbsp;    config\_args='\-dOes \-f confug\.sh'
&nbsp;    hint=previous\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define 
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, optimize='\-O'\, gccversion=2\.7\.2\.3
&nbsp;    cppflags='\-I/usr/local/include \-I/opt/local/include \-DREG\_INFTY=22786'
&nbsp;    ccflags ='\-I/usr/local/include \-I/opt/local/include \-DREG\_INFTY=22786'
&nbsp;    stdchar='unsigned char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
&nbsp;    alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags =' \-L/usr/local/lib \-L/opt/local/lib'
&nbsp;    libpth=/usr/local/lib /opt/local/lib /lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-ldl \-lm \-lc \-lcrypt
&nbsp;    libc=/lib/libc\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-G \-L/usr/local/lib \-L/opt/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Locally applied patches&#8203;:
&nbsp;      diag\.format
&nbsp;      perldoc\.vers
&nbsp;      installman\.opt
&nbsp;      undef\.opcode
&nbsp;  Built under solaris
&nbsp;  Compiled at Apr  4 2000 17&#8203;:17&#8203;:26
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/mjtg/perl5\.6\.0/lib
&nbsp;    /home/mjtg/perl5\.6\.0/lib
&nbsp;    /home/mjtg/perl5\.6\.0/lib
&nbsp;    /home/mjtg/perl5\.6\.0/lib
&nbsp;    \.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

"M\.J\.T\. Guy" wrote&#8203;:
> 
> Playing around with one of Barrie Slaymaker's bits of code\, I tried
> 
> perl5\.6\.0 \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq "fooba"\)\, @&#8203;bar'
> 
> and got
> 
> Not enough arguments for grep at \-e line 1\, near ""fooba"\)"

I trip over that\, too\.

> But the rest of the junk suggests that perhaps CHECK routines shouldn't
> be called after syntax errors\.

Not sure if this is related\, but the CHECK bit rang a bell\.

\[barries@&#8203;jester /tmp\]$ perl \-w \-MO=Deparse \-MBenchmark \-e 1
Exiting subroutine via next at /usr/local/lib/perl5/5\.6\.0/i686\-linux/B/Deparse\.pm line 257\.
Exiting subroutine via next at /usr/local/lib/perl5/5\.6\.0/i686\-linux/B/Deparse\.pm line 257\.
Can't call method "sibling" on an undefined value at /usr/local/lib/perl5/5\.6\.0/i686\-linux/B/Deparse\.pm line 257\.
CHECK failed\-\-call queue aborted\.

\- Barrie

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Barrie Slaymaker \(lists\.p5p\)&#8203;:
>\[barries@&#8203;jester /tmp\]$ perl \-w \-MO=Deparse \-MBenchmark \-e 1

Fixed in cfgperl\.

>> perl5\.6\.0 \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq "fooba"\)\, @&#8203;bar'

Not fixed in cfgperl\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Barrie Slaymaker \<barries@&#8203;slaysys\.com> wrote
> \[barries@&#8203;jester /tmp\]$ perl \-w \-MO=Deparse \-MBenchmark \-e 1
> Exiting subroutine via next at /usr/local/lib/perl5/5\.6\.0/i686\-linux/B/Deparse\.pm line 257\.
> Exiting subroutine via next at /usr/local/lib/perl5/5\.6\.0/i686\-linux/B/Deparse\.pm line 257\.
> Can't call method "sibling" on an undefined value at /usr/local/lib/perl5/5\.6\.0/i686\-linux/B/Deparse\.pm line 257\.
> CHECK failed\-\-call queue aborted\.

That's mended in perl\-current\, so is a different problem from my example\.

Tho' I note in passing that it deparses the whole of Benchmark\.pm\,
rather than just '1' as I would have expected\.    Is that doing the
Right Thing?


Mike Guy

**Comment by p5pRT** [comment]
### From @smpeters

> \[mjtg@&#8203;cus\.cam\.ac\.uk \- Mon Jun 26 00&#8203;:48&#8203;:22 2000\]&#8203;:
> 
> Playing around with one of Barrie Slaymaker's bits of code\, I tried
> 
> perl5\.6\.0 \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq
> "fooba"\)\, @&#8203;bar'
> 
> and got
> 
> Not enough arguments for grep at \-e line 1\, near ""fooba"\)"
> \-e had compilation errors\.
> unexpected OP\_GREPSTART at /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line
> 636
>     B&#8203;::Deparse&#8203;::pp\_grepstart\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::LISTOP=SCALAR\(0x16d498\)'\, 7\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
>     B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x16d498\)'\, 7\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1634
>     B&#8203;::Deparse&#8203;::pp\_list\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1810
>     B&#8203;::Deparse&#8203;::pp\_null\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
>     B&#8203;::Deparse&#8203;::deparse\(undef\, 'B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1239
>     B&#8203;::Deparse&#8203;::deparse\_binop\_right\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 'B&#8203;::UNOP=SCALAR\(0x16d954\)'\, 7\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1257
>     B&#8203;::Deparse&#8203;::binop\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 6\, '='\, 7\, 1\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1303
>     B&#8203;::Deparse&#8203;::pp\_aassign\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 6\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
>     B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::BINOP=SCALAR\(0x29c9fc\)'\, 6\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 1646
>     B&#8203;::Deparse&#8203;::pp\_list\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::LISTOP=SCALAR\(0x16d9f0\)'\, 0\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
>     B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::Deparse=HASH\(0xea960\)'\, 'B&#8203;::LISTOP=SCALAR\(0x16d9f0\)'\, 0\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 706
>     B&#8203;::Deparse&#8203;::pp\_leave\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::LISTOP=SCALAR\(0x177a5c\)'\, 0\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 398
>     B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xea960\)'\,
> 'B&#8203;::LISTOP=SCALAR\(0x177a5c\)'\, 0\) called at
> /home/mjtg/perl5\.6\.0/lib/B/Deparse\.pm line 375
>     B&#8203;::Deparse&#8203;::\_\_ANON\_\_\(\) called at \(eval 4\) line 1
>     O&#8203;::CHECK\(\) called at \-e line 1
>     require 0 called at \-e line 1
> @&#8203;x = XXX\, @&#8203;bar;
> 
> The syntax error is of course correct \(though it makes me wonder if
> the
> "if it looks like a function \.\.\. " rule isn't overagressive
> sometimes\)\.
> 
> But the rest of the junk suggests that perhaps CHECK routines
> shouldn't
> be called after syntax errors\.
> 
> 
> Mike Guy
> 


This appears to have been fixed at some point in the past\.

> perl \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq "fooba"\)\, @&#8203;bar' 
Not enough arguments for grep at \-e line 1\, near ""fooba"\)"
\-e had compilation errors\.
BEGIN \{ $^W = 1; \}
unexpected OP\_GREPSTART at
/usr/libdata/perl5/i386\-openbsd/5\.8\.5/B/Deparse\.pm line 1068\.
@&#8203;x = XXX\, @&#8203;bar;

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From mjtg@cam.ac.uk

"Steve Peters wrote
> This appears to have been fixed at some point in the past\.
>
> > perl \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq "fooba"\)\, @&#8203;bar' 
> Not enough arguments for grep at \-e line 1\, near ""fooba"\)"
> \-e had compilation errors\.
> BEGIN \{ $^W = 1; \}
> unexpected OP\_GREPSTART at
> /usr/libdata/perl5/i386\-openbsd/5\.8\.5/B/Deparse\.pm line 1068\.
> @&#8203;x = XXX\, @&#8203;bar;

It's been changed\, but it hasn't been fixed  \-  where do you think that
"unexpected OP\_GREPSTART" is coming from if it's not the CHECK routine?

This bug should be re\-opened\.


Mike Guy

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @smpeters

> \[mjtg@&#8203;cam\.ac\.uk \- Mon Nov 22 10&#8203;:18&#8203;:24 2004\]&#8203;:
> 
> "Steve Peters wrote
> > This appears to have been fixed at some point in the past\.
> >
> > > perl \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq "fooba"\)\,
> @&#8203;bar'
> > Not enough arguments for grep at \-e line 1\, near ""fooba"\)"
> > \-e had compilation errors\.
> > BEGIN \{ $^W = 1; \}
> > unexpected OP\_GREPSTART at
> > /usr/libdata/perl5/i386\-openbsd/5\.8\.5/B/Deparse\.pm line 1068\.
> > @&#8203;x = XXX\, @&#8203;bar;
> 
> It's been changed\, but it hasn't been fixed  \-  where do you think
> that
> "unexpected OP\_GREPSTART" is coming from if it's not the CHECK
> routine?
> 
> This bug should be re\-opened\.
Reopened\. :\)

I attemted to reattempt to recreate the problem and found that it only
complains loudly after adding a "\-d"\.  I also found that map\(\) has a
similar problem\.

bash\-2\.05b$ perl \-d \-MO=Deparse \-e'grep '

Loading DB routines from perl5db\.pl version 1\.27
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

Not enough arguments for grep at \-e line 1\, at EOF
\-e had compilation errors\.
 at \-e line 1
unexpected OP\_GREPSTART at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Deparse\.pm line 1068\.
 at /usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Deparse\.pm line 1068
&nbsp;        B&#8203;::Deparse&#8203;::AUTOLOAD\('B&#8203;::Deparse=HASH\(0xa3f0684\)'\,
'B&#8203;::LISTOP=SCALAR\(0xa
54ceb4\)'\, 0\.5\) called at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Depars
e\.pm line 791
&nbsp;        B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xa3f0684\)'\,
'B&#8203;::LISTOP=SCALAR\(0xa5
4ceb4\)'\, 0\.5\) called at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Deparse
\.pm line 1205
&nbsp;        B&#8203;::Deparse&#8203;::deparse\_root\('B&#8203;::Deparse=HASH\(0xa3f0684\)'\,
'B&#8203;::LISTOP=SCALAR
\(0xa54d510\)'\) called at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Deparse
\.pm line 649
&nbsp;       
B&#8203;::Deparse&#8203;::\_\_ANON\_\_\[/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/De
parse\.pm&#8203;:667\]\(\) called at \(eval
4\)\[/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64in
t/O\.pm&#8203;:22\] line 31
&nbsp;        O&#8203;::CHECK\(\) called at \-e line 1
&nbsp;        eval \{\.\.\.\} called at \-e line 1
XXX;
bash\-2\.05b$ perl \-d \-MO=Deparse \-e'map '

Loading DB routines from perl5db\.pl version 1\.27
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

Not enough arguments for map at \-e line 1\, at EOF
\-e had compilation errors\.
 at \-e line 1
unexpected OP\_MAPSTART at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Depar
se\.pm line 1068\.
 at /usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Deparse\.pm line 1068
&nbsp;        B&#8203;::Deparse&#8203;::AUTOLOAD\('B&#8203;::Deparse=HASH\(0xa3f0684\)'\,
'B&#8203;::LISTOP=SCALAR\(0xa
54ceb4\)'\, 0\.5\) called at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Depars
e\.pm line 791
&nbsp;        B&#8203;::Deparse&#8203;::deparse\('B&#8203;::Deparse=HASH\(0xa3f0684\)'\,
'B&#8203;::LISTOP=SCALAR\(0xa5
4ceb4\)'\, 0\.5\) called at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Deparse
\.pm line 1205
&nbsp;        B&#8203;::Deparse&#8203;::deparse\_root\('B&#8203;::Deparse=HASH\(0xa3f0684\)'\,
'B&#8203;::LISTOP=SCALAR
\(0xa54d510\)'\) called at
/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/Deparse
\.pm line 649
&nbsp;       
B&#8203;::Deparse&#8203;::\_\_ANON\_\_\[/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64int/B/De
parse\.pm&#8203;:667\]\(\) called at \(eval
4\)\[/usr/lib/perl5/5\.8\.5/cygwin\-thread\-multi\-64in
t/O\.pm&#8203;:22\] line 31
&nbsp;        O&#8203;::CHECK\(\) called at \-e line 1
&nbsp;        eval \{\.\.\.\} called at \-e line 1
XXX;

**Comment by p5pRT** [comment]
### From @smpeters

> \[stmpeters \- Mon Nov 22 12&#8203;:25&#8203;:14 2004\]&#8203;:
> 
> > \[mjtg@&#8203;cam\.ac\.uk \- Mon Nov 22 10&#8203;:18&#8203;:24 2004\]&#8203;:
> > 
> > "Steve Peters wrote
> > > This appears to have been fixed at some point in the past\.
> > >
> > > > perl \-w \-MO=Deparse \-e '@&#8203;x = grep \(substr\( $\_\, 0\, 5 \) eq "fooba"\)\,
> > @&#8203;bar'
> > > Not enough arguments for grep at \-e line 1\, near ""fooba"\)"
> > > \-e had compilation errors\.
> > > BEGIN \{ $^W = 1; \}
> > > unexpected OP\_GREPSTART at
> > > /usr/libdata/perl5/i386\-openbsd/5\.8\.5/B/Deparse\.pm line 1068\.
> > > @&#8203;x = XXX\, @&#8203;bar;
> > 
> > It's been changed\, but it hasn't been fixed  \-  where do you think
> > that
> > "unexpected OP\_GREPSTART" is coming from if it's not the CHECK
> > routine?
> > 
> > This bug should be re\-opened\.
> Reopened\. :\)
> 
> I attemted to reattempt to recreate the problem and found that it only
> complains loudly after adding a "\-d"\.  I also found that map\(\) has a
> similar problem\.
> 
> bash\-2\.05b$ perl \-d \-MO=Deparse \-e'grep '
\<snip>

The error was caused by pp\_grepstart\(\) and pp\_mapstart\(\) not being
implemented in B&#8203;::Deparse\.  This caused the errors generated by
B&#8203;::Deparse&#8203;::AUTOLOAD\(\) to be generated\.  The attached patch fixes this bug\.

The patch was built from blead\-23523\.  All tests were successful\.

It also behaves as other built\-in functions do\.

> \./perl \-Ilib \-d \-MO=Deparse \-e'rindex'

Loading DB routines from perl5db\.pl version 1\.28
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

Not enough arguments for rindex at \-e line 1\, at EOF
\-e had compilation errors\.
 at \-e line 1
rindex;
> \./perl \-Ilib \-d \-MO=Deparse \-e'grep'

Loading DB routines from perl5db\.pl version 1\.28
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

Not enough arguments for grep at \-e line 1\, at EOF
\-e had compilation errors\.
 at \-e line 1
grep;

**Comment by p5pRT** [comment]
### From @smpeters

<details><summary>B_Deparse_grepstart.patch</summary>

```diff
--- ext/B/B/Deparse.pm.orig	Mon Nov 22 15:10:30 2004
+++ ext/B/B/Deparse.pm	Mon Nov 22 15:12:35 2004
@@ -2366,6 +2366,8 @@
 
 sub pp_mapwhile { mapop(@_, "map") }
 sub pp_grepwhile { mapop(@_, "grep") }
+sub pp_mapstart { baseop(@_, "map") }
+sub pp_grepstart { baseop(@_, "grep") }
 
 sub pp_list {
     my $self = shift;

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Steve Peters via RT wrote&#8203;:
> The error was caused by pp\_grepstart\(\) and pp\_mapstart\(\) not being
> implemented in B&#8203;::Deparse\.  This caused the errors generated by
> B&#8203;::Deparse&#8203;::AUTOLOAD\(\) to be generated\.  The attached patch fixes this bug\.

Thanks\, applied as \#23526\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
